File: examples/C/wuclient.c
//
//  Weather update client
//  Connects SUB socket to tcp://localhost:5556
//  Collects weather updates and finds avg temp in zipcode
//
#include "zhelpers.h"

int main (int argc, char *argv [])
{
    void *context = zmq_init (1);

    //  Socket to talk to server
    printf ("Collecting updates from weather server...\n");
    void *subscriber = zmq_socket (context, ZMQ_SUB);
    zmq_connect (subscriber, "tcp://localhost:5556");

    //  Subscribe to zipcode, default is NYC, 10001
    char *filter = (argc > 1)? argv [1]: "10001 ";
    zmq_setsockopt (subscriber, ZMQ_SUBSCRIBE, filter, strlen (filter));

    //  Process 100 updates
    int update_nbr;
    long total_temp = 0;
    for (update_nbr = 0; update_nbr < 100; update_nbr++) {
        char *string = s_recv (subscriber);
        int zipcode, temperature, relhumidity;
        sscanf (string, "%d %d %d",
            &zipcode, &temperature, &relhumidity);
        total_temp += temperature;
        free (string);
    }
    printf ("Average temperature for zipcode '%s' was %dF\n",
        filter, (int) (total_temp / update_nbr));

    zmq_close (subscriber);
    zmq_term (context);
    return 0;
}

File: examples/OCaml/wuclient.ml
(**
  Weather update client
  Connects SUB socket to tcp://localhost:5556
  Collects weather updates and finds avg temp in zipcode
*)

open Zmq
open Helpers

let () =
  with_context @@ fun ctx ->
  with_socket ctx Socket.sub @@ fun sub ->
  printfn "Collecting updates from weather server...";
  Socket.connect sub "tcp://localhost:5556";
  (* Subscribe to zipcode, default is NYC, 10001 *)
  let filter = match Array.to_list Sys.argv with _::zip::_ -> zip | _ -> "10001 " in
  Socket.subscribe sub filter;
  (* Process 100 updates *)
  let total_temp = ref 0 in
  for i = 0 to pred 100 do
    Scanf.sscanf (Socket.recv sub) "%d %d %d" begin fun _zipcode temperature _relhumidity ->
      total_temp := !total_temp + temperature
    end
  done;
  printfn "Average temperature for zipcode '%s' was %dF" filter (!total_temp / 100)
